File: oauth2-java-backend/src/main/java/com/example/auth/memory/InMemoryResourceOwnerRepository.java
package com.example.auth.memory;

import com.clouway.oauth2.ResourceOwner;
import com.clouway.oauth2.ResourceOwnerAuthentication;
import com.clouway.oauth2.ResourceOwnerStore;
import com.clouway.oauth2.Session;
import com.clouway.oauth2.SessionSecurity;
import com.clouway.oauth2.token.TokenGenerator;
import com.google.common.base.Optional;
import com.google.inject.Inject;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * @author Ivan Stefanov <ivan.stefanov@clouway.com>
 */
class InMemoryResourceOwnerRepository implements ResourceOwnerStore, ResourceOwnerAuthentication, SessionSecurity {
  private Map<String, ResourceOwner> resourceOwners = new HashMap<String, ResourceOwner>() {{
    put("admin", new ResourceOwner("admin", "admin"));
  }};

  private Set<Session> sessions = new HashSet<Session>();

  private TokenGenerator tokenGenerator;

  @Inject
  InMemoryResourceOwnerRepository(TokenGenerator tokenGenerator) {
    this.tokenGenerator = tokenGenerator;
  }

  @Override
  public void save(ResourceOwner resourceOwner) {
    resourceOwners.put(resourceOwner.username, resourceOwner);
  }

  @Override
  public Optional<Session> auth(String username, String password, String remoteAddress) {
    if (resourceOwners.containsKey(username)) {
      if (password.equals(resourceOwners.get(username).password)) {
        Session session = new Session(tokenGenerator.generate());

        sessions.add(session);

        return Optional.of(session);
      }
    }

    return Optional.absent();
  }

  @Override
  public Boolean exists(Session session) {
    return sessions.contains(session);
  }
}

File: oauth2-example-app/src/main/java/com/clouway/oauth2/exampleapp/storage/InMemoryResourceOwnerRepository.kt
package com.clouway.oauth2.exampleapp.storage

import com.clouway.oauth2.Session
import com.clouway.oauth2.exampleapp.ResourceOwner
import com.clouway.oauth2.exampleapp.ResourceOwnerAuthentication
import com.clouway.oauth2.exampleapp.ResourceOwnerStore
import com.clouway.oauth2.exampleapp.SessionSecurity
import com.clouway.oauth2.token.TokenGenerator
import com.google.common.base.Optional
import com.google.inject.Inject

/**
 * @author Ivan Stefanov <ivan.stefanov></ivan.stefanov>@clouway.com>
 */
internal class InMemoryResourceOwnerRepository
    @Inject
    constructor(
        private val tokenGenerator: TokenGenerator,
    ) : ResourceOwnerStore,
        ResourceOwnerAuthentication,
        SessionSecurity {
        private val resourceOwners = mutableMapOf("admin" to ResourceOwner("admin", "admin"))

        private val sessions: MutableSet<Session> = HashSet()

        override fun save(resourceOwner: ResourceOwner) {
            resourceOwners[resourceOwner.username] = resourceOwner
        }

        override fun auth(
            username: String,
            password: String,
            remoteAddress: String,
        ): Optional<Session> {
            if (resourceOwners.containsKey(username)) {
                if (password == resourceOwners[username]!!.password) {
                    val session = Session(tokenGenerator.generate())
				
                    sessions.add(session)
				
                    return Optional.of(session)
                }
            }
		
            return Optional.absent()
        }

        override fun exists(session: Session): Boolean = sessions.contains(session)
    }
